File: Sesion 12/Ejercicio/Ejercicio1 S12/EJERCICIO1 S12.py
class Inventory:
    def __init__(self):
        self.products = {}

    def add_product(self, code, description, quantity):
        self.products[code] = {'description': description, 'quantity': quantity}

    def update_quantity(self, code, quantity):
        if code in self.products:
            self.products[code]['quantity'] = quantity
        else:
            print("Product not found!")

    def search_product(self, code):
        if code in self.products:
            product = self.products[code]
            print(f"Product found: {product['description']} - Quantity: {product['quantity']}")
        else:
            print("Product not found!")

    def list_products(self):
        sorted_products = sorted(self.products.values(), key=lambda x: x['description'])
        print("Product List:")
        for product in sorted_products:
            print(f"{product['description']} - Code: {self.get_code_by_description(product['description'])} - Quantity: {product['quantity']}")

    def get_code_by_description(self, description):
        for code, product in self.products.items():
            if product['description'] == description:
                return code

def display_menu():
    print("Inventory Management System")
    print("1. Add Product")
    print("2. Update Product Quantity")
    print("3. Search Product")
    print("4. List Products")
    print("5. Exit")
    return int(input("Choose an option: "))

def main():
    inventory = Inventory()
    choice = 0

    while choice != 5:
        choice = display_menu()
        
        if choice == 1:
            code = input("Enter product code: ")
            description = input("Enter product description: ")
            quantity = int(input("Enter product quantity: "))
            inventory.add_product(code, description, quantity)
        elif choice == 2:
            code = input("Enter product code: ")
            quantity = int(input("Enter new quantity: "))
            inventory.update_quantity(code, quantity)
        elif choice == 3:
            code = input("Enter product code: ")
            inventory.search_product(code)
        elif choice == 4:
            inventory.list_products()
        elif choice == 5:
            print("Exiting...")
        else:
            print("Invalid choice! Please try again.")
        print()

if __name__ == "__main__":
    main()

File: Sesion 12/Ejercicio/Ejercicio1 S12/EJERCICIO1 S12.cpp
#include <iostream>
#include <string>
#include <map>
#include <vector>
#include <algorithm>

using namespace std;

class Inventory {
private:
    struct Product {
        string code;
        string description;
        int quantity;
    };

    map<string, Product> products;

public:
    void addProduct(const string& code, const string& description, int quantity) {
        products[code] = { code, description, quantity };
    }

    void updateQuantity(const string& code, int quantity) {
        if (products.find(code) != products.end()) {
            products[code].quantity = quantity;
        } else {
            cout << "Product not found!" << endl;
        }
    }

    void searchProduct(const string& code) {
        auto it = products.find(code);
        if (it != products.end()) {
            cout << "Product found: " << it->second.description << " - Quantity: " << it->second.quantity << endl;
        } else {
            cout << "Product not found!" << endl;
        }
    }

    void listProducts() {
        vector<Product> productList;
        for (const auto& pair : products) {
            productList.push_back(pair.second);
        }

        sort(productList.begin(), productList.end(), [](const Product& a, const Product& b) {
            return a.description < b.description;
        });

        cout << "Product List:" << endl;
        for (const auto& product : productList) {
            cout << product.description << " - Code: " << product.code << " - Quantity: " << product.quantity << endl;
        }
    }
};

void displayMenu() {
    cout << "Inventory Management System" << endl;
    cout << "1. Add Product" << endl;
    cout << "2. Update Product Quantity" << endl;
    cout << "3. Search Product" << endl;
    cout << "4. List Products" << endl;
    cout << "5. Exit" << endl;
    cout << "Choose an option: ";
}

int main() {
    Inventory inventory;
    int choice;
    string code, description;
    int quantity;

    do {
        displayMenu();
        cin >> choice;

        switch (choice) {
            case 1:
                cout << "Enter product code: ";
                cin >> code;
                cin.ignore(); 
                cout << "Enter product description: ";
                getline(cin, description);
                cout << "Enter product quantity: ";
                cin >> quantity;
                inventory.addProduct(code, description, quantity);
                break;
            case 2:
                cout << "Enter product code: ";
                cin >> code;
                cout << "Enter new quantity: ";
                cin >> quantity;
                inventory.updateQuantity(code, quantity);
                break;
            case 3:
                cout << "Enter product code: ";
                cin >> code;
                inventory.searchProduct(code);
                break;
            case 4:
                inventory.listProducts();
                break;
            case 5:
                cout << "Exiting..." << endl;
                break;
            default:
                cout << "Invalid choice! Please try again." << endl;
                break;
        }

        cout << endl;
    } while (choice != 5);

    return 0;
}
